pubsub/pubsub.go

```diff
@@ -15,32 +15,34 @@ type Message interface{}
 // Publish publishes a message to a channel.
 func Publish(channel string, data Message) {
 	mu.Lock()
+	defer mu.Unlock()
 	for _, listener := range listeners[channel] {
-		if listener != nil {
-			listener <- data
-		}
+		listener <- data
 	}
-	mu.Unlock()
 }
 
 // Subscribe subscribes to the channel and returns listener for it.
 func Subscribe(channel string) chan Message {
-	listener := make(chan Message)
 	mu.Lock()
+	defer mu.Unlock()
+	listener := make(chan Message)
 	if listeners[channel] == nil {
 		listeners[channel] = make([]chan Message, 0)
 	}
 	listeners[channel] = append(listeners[channel], listener)
-	mu.Unlock()
 	return listener
 }
 
-// Unsubscribe unsubscribes a subscription for listening channel.
-func Unsubscribe(channel string, subscription chan Message) {
+// Unsubscribe unsubscribes a listener from listening channel.
+func Unsubscribe(channel string, listener chan Message) {
 	mu.Lock()
 	defer mu.Unlock()
-	for i, s := range listeners[channel] {
-		if s == subscription {
+	if len(listeners[channel]) == 1 {
```

Comment:
I'm not sure about that one. What happen if for any reason (by design it shouldn't be possible but everything is possible) the list of `listeners[channel]` contains another `chan` than `listener`.

I would prefer to iterate over the `listeners` and if it's just about memory optimization we can still at the end of the loop, if the length of `listeners[channel]` is 0 then delete it

Is there any other reason why you did this condition here ?